nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/util/SecureNiFiConfigUtil.java

```diff
@@ -0,0 +1,197 @@
+import java.security.GeneralSecurityException;
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class SecureNiFiConfigUtil {
+
+    private static final int CERT_DURATION_DAYS = 60;
+    private static final String LOCALHOST_NAME = "localhost";
+    private static final String PROPERTY_VALUE_PATTERN = "%s=%s";
+
+    private SecureNiFiConfigUtil() {
+
+    }
+
+    private static boolean fileExists(String filename) {
+        return StringUtils.isNotEmpty(filename) && Paths.get(filename).toFile().exists();
+    }
+
+    private static boolean isHttpsSecurityConfigured(final Properties nifiProperties) {
+        if (StringUtils.isEmpty(nifiProperties.getProperty(NiFiProperties.WEB_HTTPS_PORT, StringUtils.EMPTY))) {
+            return false;
+        }
+
+        String keystorePath = nifiProperties.getProperty(NiFiProperties.SECURITY_KEYSTORE, StringUtils.EMPTY);
+        String truststorePath = nifiProperties.getProperty(NiFiProperties.SECURITY_TRUSTSTORE, StringUtils.EMPTY);
+        if (StringUtils.isEmpty(keystorePath) || StringUtils.isEmpty(truststorePath)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * If HTTPS is enabled (nifi.web.https.port is set), but the keystore file specified in nifi.security.keystore
+     * does not exist, this will generate a key pair and self-signed certificate, generate the associated keystore
+     * and truststore and write them to disk under the configured filepaths, generate a secure random keystore password
+     * and truststore password, and write these to the nifi.properties file.
+     * @param nifiPropertiesFilename The filename of the nifi.properties file
+     * @param cmdLogger The bootstrap logger
+     * @throws IOException can be thrown when writing keystores to disk
+     * @throws RuntimeException indicates a security exception while generating keystores
+     */
+    public static void configureSecureNiFiProperties(String nifiPropertiesFilename, Logger cmdLogger) throws IOException, RuntimeException {
+        final File propertiesFile = new File(nifiPropertiesFilename);
+        final Properties nifiProperties = loadProperties(propertiesFile);
+
+        if (!isHttpsSecurityConfigured(nifiProperties)) {
+            cmdLogger.debug("No HTTPS configuration detected: skipping Apache Nifi certificate generation.");
+            return;
+        }
+
+        String keystorePath = nifiProperties.getProperty(NiFiProperties.SECURITY_KEYSTORE, StringUtils.EMPTY);
+        String truststorePath = nifiProperties.getProperty(NiFiProperties.SECURITY_TRUSTSTORE, StringUtils.EMPTY);
+        boolean keystoreExists = fileExists(keystorePath);
+        boolean truststoreExists = fileExists(truststorePath);
+
+        if (!keystoreExists && !truststoreExists) {
```

Comment:
@gresockj @exceptionfactory  Do you think we'll run into cases where someone was intending to use their own keystore/truststore, but accidentally put the wrong path or something, so it ends up looking like they didn't exist and we proceed to generate stuff, the app starts up fine, and they have no idea its actually not using their own? 

The only options I could think of to protect against this are...

1) Adding an explicit property like `nifi.security.tls.auto.generate` (or some better name), that would default to false, and they'd have to specifically enable it, and then if its enabled and the files exist already, it'd be a failure.

2) We could so that we only do the auto-generation when only the two file paths are populated, meaning the passwords and types also need to be blank, and the files don't exist, making it a stronger indicator that they wanted auto generated certs. It loses the ability to provide your own passwords, but maybe doesn't matter if this is intended for helping with local default security?

What do you guys think?

Reply:
Good thought, Bryan.  I'm leaning toward #2, which is what the use case should be anyway.  Thoughts from @exceptionfactory?

Reply:
I agree that this could be an issue.  Although Option 1 is more explicit, Option 2 avoids an additional property while also adding more confirmation around this scenario.  Creating a new method that wraps the check for both keystore and truststore files existing, as well as the absence of passwords, seems like a good way to go.